aten/src/ATen/native/cpu/layer_norm_kernel.cpp

```diff
@@ -358,8 +357,274 @@ void LayerNormBackwardKernelImplInternal(
+  BFloat16* buffer_data = nullptr;
+  if (!dgamma_null || !dbeta_null) {
+    // zero the immediate buffer and skip zero dgamma and dbeta
+    buffer.resize_({2, num_threads, N}).zero_();
+    buffer_data = buffer.template data_ptr<BFloat16>();
+  }
+
+  // First path of dgamma/dbeta and dX
+  at::parallel_for(0, M, 1, [&](int64_t start, int64_t end) {
+    int tid = at::get_thread_num();
+    TORCH_CHECK(
+        tid < num_threads,
+        "expect thread id smaller than ",
+        num_threads,
+        ", got thread id ",
+        tid);
+    BFloat16* dgamma_buffer_ptr = dgamma_null ? nullptr : buffer_data + tid * N;
+    BFloat16* dbeta_buffer_ptr =
+        dbeta_null ? nullptr : buffer_data + num_threads * N + tid * N;
+    for (const auto i : c10::irange(start, end)) {
+      const BFloat16* dY_ptr = dY_data + i * N;
+      const BFloat16* X_ptr = X_data + i * N;
+      if (!dgamma_null) {
+        const float a = rstd_data[i];
+        const float b = -a * mean_data[i];
+        // Scalar math:
+        // for (const auto j : c10::irange(N)) {
+        //   dgamma_data[j] += dY_ptr[j] * (a * X_ptr[j] + b);
+        // }
+        vec::map3<BFloat16>(
+            [a, b](fVec dgamma, fVec dy, fVec x) {
+              return dgamma + dy * (fVec(a) * x + fVec(b));
+            },
+            dgamma_buffer_ptr,
+            dgamma_buffer_ptr,
+            dY_ptr,
+            X_ptr,
+            N);
+      }
+      if (!dbeta_null) {
+        // Scalar math:
+        // for (const auto j : c10::irange(N)) {
+        //   dbeta_data[j] += dY_ptr[j];
+        // }
+        vec::map2<BFloat16>(
+            [](fVec dbeta, fVec dy) { return dbeta + dy; },
+            dbeta_buffer_ptr,
+            dbeta_buffer_ptr,
+            dY_ptr,
+            N);
+      }
+      if (!dX_null) {
+        BFloat16* dX_ptr = dX_data + i * N;
+        float ds = float(0);
+        float db = float(0);
+        // Scalar math:
+        // for (const auto j : c10::irange(N)) {
+        //   const T gamma_v = gamma_null ? T(1) : gamma_data[j];
+        //   ds += dY_ptr[j] * X_ptr[j] * gamma_v;
+        //   db += dY_ptr[j] * gamma_v;
+        // }
+        if (gamma_null) {
+          ds = vec::map2_reduce_all<BFloat16>(
+              [](fVec x, fVec y) { return x * y; },
+              [](fVec x, fVec y) { return x + y; },
+              dY_ptr,
+              X_ptr,
+              N);
+          db = vec::reduce_all<BFloat16>(
+              [](fVec& x, fVec& y) { return x + y; }, dY_ptr, N);
+        } else {
+          if (N < bVec::size()) {
+            bVec x_bvec = bVec::loadu(X_ptr, N);
+            bVec dy_bvec = bVec::loadu(dY_ptr, N);
+            fVec x_fvec0, x_fvec1, dy_fvec0, dy_fvec1, gamma_fvec0, gamma_fvec1;
+            std::tie(x_fvec0, x_fvec1) = convert_bfloat16_float(x_bvec);
+            std::tie(dy_fvec0, dy_fvec1) = convert_bfloat16_float(dy_bvec);
+            std::tie(gamma_fvec0, gamma_fvec1) = load2f(gamma_data, N);
```

Comment:
> Looks like `layer_norm_backward_kernel_mixed_type` is just a copy of `LayerNormBackwardKernelImplInternal` with slightly different argument types.
> 
> Perhaps a better approach would be to just move `ACC_T` as template argument of `LayerNormBackwardKernelImplInternal` and call it slightly differently in mixed precision type scenario

This backward implementation is consistent with forward, and actually, there are many places that need to be modified to support mixed data type, not just slightly different argument type. For example, here, we cannot simply call vec::map3_reduce_all to do this part. Thanks.